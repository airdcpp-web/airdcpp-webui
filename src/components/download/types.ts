import * as UI from 'types/ui';


export type BrowseHandler = (() => void) | undefined;

export type PathDownloadHandler = (targetPath: string, targetFilename?: string) => Promise<any>;

export type LayoutProps = React.PropsWithChildren<{
  menuItems: React.ReactNode[];
  title: string;
  handleBrowse: BrowseHandler;
  t: UI.TranslateF;
}>;
