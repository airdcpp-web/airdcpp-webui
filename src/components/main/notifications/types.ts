export type NotificationLevel = 'error' | 'warning' | 'info' | 'success';
