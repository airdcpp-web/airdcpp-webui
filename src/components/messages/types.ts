import { IntersectionObserverProps } from 'react-intersection-observer';

import * as UI from '@/types/ui';

export interface CommonMessageProps extends Omit<IntersectionObserverProps, 'children'> {
  highlightMenuProps: UI.MessageActionMenuData;
}
