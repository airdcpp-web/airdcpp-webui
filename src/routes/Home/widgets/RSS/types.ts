import * as UI from '@/types/ui';

export type FeedNode<Attributes extends object = UI.EmptyObject> =
  | string
  | {
      text: string;
      attr: Attributes;
    };

export interface FeedItem {
  title?: FeedNode;
  updated?: FeedNode;
  pubDate?: FeedNode;
  guid?: FeedNode;
  id?: FeedNode;
  link?: FeedNode<{
    href: string;
  }>;
}

export interface RawFeedData {
  rss?: {
    channel?: {
      item?: FeedItem[];
    };
  };
  feed?: {
    entry?: FeedItem[];
  };
}
