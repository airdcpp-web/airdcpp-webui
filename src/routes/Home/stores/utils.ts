import { saveWidgetSettings } from '@/routes/Home/widgets/WidgetUtils';
import { Layouts } from 'react-grid-layout';

import * as UI from '@/types/ui';
import {
  ApplicationWidgetInfo,
  ExtensionWidgetInfo,
  NotepadWidgetInfo,
  RSSWidgetInfo,
  TransferWidgetInfo,
} from '../widgets';
import { HomeLayoutColumns } from '../constants/HomeLayoutConstants';

export const createLayoutWidget = (
  layouts: Layouts,
  widgetInfo: UI.Widget,
  id: string,
  x?: number,
  y?: number,
) => {
  // Add the same widget for all layouts (we are lazy and use the same size for all layouts)
  return Object.keys(HomeLayoutColumns).reduce((reducedLayouts, key) => {
    reducedLayouts[key] = layouts[key] || [];
    reducedLayouts[key] = reducedLayouts[key].concat({
      i: id,
      x: x || (reducedLayouts[key].length * 2) % (HomeLayoutColumns[key] || 12),
      y: y || 0,
      ...widgetInfo.size,
    });

    return reducedLayouts;
  }, {} as Layouts);
};

const createDefaultLayoutWidget = <SettingsT>(
  layouts: Layouts,
  widgetInfo: UI.Widget,
  x: number,
  y: number,
  name?: string,
  settings?: SettingsT,
  suffix = '_default',
) => {
  const id = widgetInfo.typeId + suffix;
  const newLayout = createLayoutWidget(layouts, widgetInfo, id, x, y);

  saveWidgetSettings(id, {
    name,
    widget: settings,
  });

  return newLayout;
};

const createDefaultLayoutWidgets = (layouts: Layouts) => {
  layouts = createDefaultLayoutWidget(layouts, ApplicationWidgetInfo, 0, 0);
  layouts = createDefaultLayoutWidget(
    layouts,
    RSSWidgetInfo,
    ApplicationWidgetInfo.size.w,
    0,
    'News',
    {
      feed_url: 'https://airdcpp-web.github.io/feed.xml',
    },
    '_releases',
  );

  layouts = createDefaultLayoutWidget(
    layouts,
    ExtensionWidgetInfo,
    ApplicationWidgetInfo.size.w + RSSWidgetInfo.size.w,
    0,
  );
  layouts = createDefaultLayoutWidget(
    layouts,
    TransferWidgetInfo,
    ApplicationWidgetInfo.size.w + RSSWidgetInfo.size.w + ExtensionWidgetInfo.size.w,
    0,
  );
  layouts = createDefaultLayoutWidget(layouts, NotepadWidgetInfo, 0, 5);
  return layouts;
};

export const ensureDefaultLayoutWidgets = (layouts: Layouts) => {
  if (Object.keys(layouts).length !== 0) {
    return layouts;
  }

  // Initialize the default layout
  return createDefaultLayoutWidgets(layouts);
};
