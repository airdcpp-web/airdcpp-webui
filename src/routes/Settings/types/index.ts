import * as API from '@/types/api';
import * as UI from '@/types/ui';

export type SettingSectionLayoutProps = Pick<UI.RouteComponentProps, 'location'> &
  React.PropsWithChildren<{
    contentClassname: string;
    selectedRootMenuItem: RootSectionType;
    selectedChildMenuItem: ChildSectionType;
    menu: {
      childMenuItems: React.ReactNode[];
      childAdvancedMenuItems?: React.ReactNode[];
      rootMenuItems: React.ReactNode[];
    };
    settingsT: UI.ModuleTranslator;
    message: React.ReactNode;
    getSaveButton: (className?: string) => React.ReactNode;
  }>;

export type SettingPageProps = React.PropsWithChildren<{
  settingsT: UI.ModuleTranslator;
  moduleT: UI.ModuleTranslator;
}>;

export interface SectionBase {
  url: string;
  access?: API.AccessEnum;
  title: string;
}

export interface ChildSectionType extends SectionBase {
  noSave?: boolean;
  local?: boolean;
  component: React.ComponentType<SettingPageProps>;
}

export interface RootSectionType extends SectionBase {
  menuItems: ChildSectionType[];
  advancedMenuItems?: ChildSectionType[];
  icon: string;
  component: React.ComponentType;
}
