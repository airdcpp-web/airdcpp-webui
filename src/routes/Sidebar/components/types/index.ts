//import * as React from 'react';

import { IconType } from 'components/semantic/Icon';

import * as API from 'types/api';
import * as UI from 'types/ui';

export interface SessionLayoutLabels {
  // Label for button that opens a new session
  newCaption?: React.ReactNode;

  // Label for button that opens a new session
  newDescription?: React.ReactNode;

  // Label for button that opens a new session
  newIcon?: IconType;
}

export interface SessionLayoutManageProps<
  SessionT extends UI.SessionItemBase,
  SessionApiT extends object = UI.EmptyObject,
  UIActionT extends UI.ActionListType<UI.SessionItemBase> = UI.EmptyObject,
> {
  // Unique ID of the section (used for storing and loading the previously open tab)
  baseUrl: string;

  // Array of the items to list
  items: SessionT[];

  // Session actions to show in the action menu
  actionIds?: string[];

  uiActions: UI.ModuleActions<SessionT, void, UIActionT>;
  sessionApi: UI.SessionActions<SessionT> & SessionApiT;

  // Item ID that is currently active (if any)
  activeId: API.IdType | undefined;

  // AccessConstant defining whether the user has edit permission
  editAccess: API.AccessEnum;

  unreadInfoStore: UI.UnreadInfoStore;

  remoteMenuId: string;
}

interface SessionMainLayoutComponents {
  newButton: React.ReactElement<any> | null;
  listActionMenuGetter: () => React.ReactNode;
  itemHeaderTitle: React.ReactNode;
  itemHeaderDescription: React.ReactNode;
  itemHeaderIcon: IconType;
  sessionMenuItems: React.ReactNode[];
}

export interface SessionMainLayoutProps<
  SessionT extends UI.SessionItemBase,
  SessionApiT extends object = UI.EmptyObject,
  UIActionsT extends UI.ActionListType<UI.SessionItemBase> = UI.EmptyObject,
> extends Pick<
      SessionLayoutManageProps<SessionT, SessionApiT, UIActionsT>,
      'unreadInfoStore'
    >,
    SessionMainLayoutComponents {
  activeItem: SessionT | null;
  onKeyDown: (event: React.KeyboardEvent) => void;
  children: React.ReactNode;

  actions: SessionLayoutManageProps<SessionT, SessionApiT, UIActionsT>['uiActions'];
}

export type SessionChildProps<
  SessionT extends UI.SessionItemBase,
  SessionApiT extends object = UI.EmptyObject,
  UIActionsT extends UI.ActionListType<UI.SessionItemBase> = UI.EmptyObject,
> = Pick<
  SessionLayoutManageProps<SessionT, SessionApiT, UIActionsT>,
  'sessionApi' | 'uiActions'
> &
  UI.RouteComponentProps & {
    session: SessionT;
    sessionT: UI.ModuleTranslator;
  };

export interface NewSessionLayoutProps extends UI.RouteComponentProps {
  sessionT: UI.ModuleTranslator;
}

export interface SessionLocationState {
  pending?: boolean;
}
