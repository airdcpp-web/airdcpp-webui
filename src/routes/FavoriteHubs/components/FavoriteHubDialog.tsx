import { useMemo, useRef } from 'react';
import RouteModal from 'components/semantic/RouteModal';

import ModalRouteDecorator, {
  ModalRouteDecoratorChildProps,
} from 'decorators/ModalRouteDecorator';

import ShareProfileConstants from 'constants/ShareProfileConstants';
import FavoriteHubConstants from 'constants/FavoriteHubConstants';

import ShareProfileDecorator, {
  ShareProfileDecoratorChildProps,
} from 'decorators/ShareProfileDecorator';
import IconConstants from 'constants/IconConstants';

import t from 'utils/tcomb-form';

import Form, {
  FormFieldChangeHandler,
  FormFieldSettingHandler,
  FormSaveHandler,
} from 'components/form/Form';
import { normalizeEnumValue, intTransformer, translateForm } from 'utils/FormUtils';

import * as API from 'types/api';
import * as UI from 'types/ui';

import { FavoriteHubEntry } from 'types/api';
import { IndeterminateCheckboxField } from 'components/form/fields/IndeterminateCheckboxField';
import { profileToEnumValue } from 'utils/ShareProfileUtils';
import { Formatter, useFormatter } from 'context/FormatterContext';

const ConnectivityModeOptions: API.SettingEnumOption[] = [
  {
    id: API.ConnectivityModeEnum.DISABLED,
    name: 'Disabled',
  },
  {
    id: API.ConnectivityModeEnum.ACTIVE_MODE,
    name: 'Active mode',
  },
  {
    id: API.ConnectivityModeEnum.PASSIVE_MODE,
    name: 'Passive mode',
  },
];

const Fields: UI.FormFieldDefinition[] = [
  {
    key: 'generic',
    title: 'Hub properties',
    type: API.SettingTypeEnum.STRUCT,
    definitions: [
      {
        key: 'name',
        title: 'Name',
        type: API.SettingTypeEnum.STRING,
      },
      {
        key: 'hub_url',
        title: 'Hub URL',
        type: API.SettingTypeEnum.STRING,
      },
      {
        key: 'hub_description',
        title: 'Hub description',
        type: API.SettingTypeEnum.STRING,
        optional: true,
      },
      {
        key: 'share_profile',
        title: 'Share profile',
        type: API.SettingTypeEnum.NUMBER,
        // eslint-disable-next-line max-len
        help: 'Custom share profiles can be selected only after entering an ADC hub address (starting with adc:// or adcs://)',
        optional: true,
      },
      {
        key: 'auto_connect',
        title: 'Auto connect',
        type: API.SettingTypeEnum.BOOLEAN,
      },
    ],
  },
  {
    key: 'user',
    title: 'User setting overrides',
    type: API.SettingTypeEnum.STRUCT,
    definitions: [
      {
        key: 'nick',
        title: 'Nick',
        type: API.SettingTypeEnum.STRING,
        optional: true,
      },
      {
        key: 'user_description',
        title: 'User description',
        type: API.SettingTypeEnum.STRING,
        optional: true,
      },
    ],
  },
  {
    key: 'connectivityV4',
    title: 'Connectivity setting overrides (IPv4)',
    type: API.SettingTypeEnum.STRUCT,
    definitions: [
      {
        key: 'connection_mode_v4',
        title: 'Connectivity mode',
        type: API.SettingTypeEnum.NUMBER,
        optional: true,
        options: ConnectivityModeOptions,
      },
      {
        key: 'connection_ip_v4',
        title: 'IP address',
        type: API.SettingTypeEnum.STRING,
        optional: true,
      },
    ],
  },
  {
    key: 'connectivityV6',
    title: 'Connectivity setting overrides (IPv6)',
    type: API.SettingTypeEnum.STRUCT,
    definitions: [
      {
        key: 'connection_mode_v6',
        title: 'Connectivity mode',
        type: API.SettingTypeEnum.NUMBER,
        optional: true,
        options: ConnectivityModeOptions,
      },
      {
        key: 'connection_ip_v6',
        title: 'IP address',
        type: API.SettingTypeEnum.STRING,
        optional: true,
      },
    ],
  },
  {
    key: 'misc',
    title: 'Miscellaneous overrides',
    type: API.SettingTypeEnum.STRUCT,
    definitions: [
      {
        key: 'use_main_chat_notify',
        title: 'Use chat notification',
        type: API.SettingTypeEnum.BOOLEAN,
        optional: true,
      },
      {
        key: 'show_joins',
        title: 'Show joins/parts',
        type: API.SettingTypeEnum.BOOLEAN,
        optional: true,
      },
      {
        key: 'away_message',
        title: 'Away message',
        type: API.SettingTypeEnum.TEXT,
        optional: true,
      },
    ],
  },
];

interface Entry extends UI.FormValueMap {
  generic: Pick<
    FavoriteHubEntry,
    'name' | 'hub_url' | 'hub_description' | 'auto_connect' | 'share_profile'
  >;
  user: Pick<FavoriteHubEntry, 'nick' | 'user_description'>;
  connectivityV4: Pick<FavoriteHubEntry, 'connection_mode_v4' | 'connection_ip_v4'>;
  connectivityV6: Pick<FavoriteHubEntry, 'connection_mode_v6' | 'connection_ip_v6'>;
  misc: Pick<FavoriteHubEntry, 'use_main_chat_notify' | 'show_joins' | 'away_message'>;
}

const toFormEntry = (entry?: API.FavoriteHubEntry): Entry | undefined => {
  if (!entry) {
    return undefined;
  }

  return {
    generic: entry,
    user: entry,
    connectivityV4: entry,
    connectivityV6: entry,
    misc: entry,
  };
};

const toFavoriteHub = (
  entry: RecursivePartial<Entry>,
): RecursivePartial<API.FavoriteHubEntry> => {
  return {
    ...entry.generic,
    ...entry.user,
    ...entry.connectivityV4,
    ...entry.connectivityV6,
    ...entry.misc,
  };
};

const isAdcHub = (hubUrl?: string) =>
  !!hubUrl && (hubUrl.startsWith('adc://') || hubUrl.startsWith('adcs://'));

// Get selection values for the profiles field
const getFieldProfiles = (
  profiles: API.ShareProfile[],
  url: string | undefined,
  formatter: Formatter,
): UI.FormOption[] => {
  return profiles
    .filter((p) => isAdcHub(url) || p.id === ShareProfileConstants.HIDDEN_PROFILE_ID)
    .map((p) => profileToEnumValue(p, formatter))
    .map(normalizeEnumValue);
};

interface FavoriteHubDialogProps {
  favT: UI.ModuleTranslator;
}

interface DataProps {
  hubEntry?: API.FavoriteHubEntry;
}

type Props = FavoriteHubDialogProps &
  DataProps &
  ShareProfileDecoratorChildProps &
  ModalRouteDecoratorChildProps;

const FavoriteHubDialog: React.FC<Props> = ({
  hubEntry,
  location,
  favT,
  profiles,
  socket,
}) => {
  const formatter = useFormatter();

  const formValue = useMemo<Entry | undefined>(() => toFormEntry(hubEntry), []);
  const formDefinitions = useMemo<UI.FormFieldDefinition[]>(
    () => translateForm(Fields, favT),
    [],
  );
  const nullOption = useMemo<UI.FormOption<string>>(
    () => ({
      value: 'null',
      text: favT.translate('Global default'),
    }),
    [],
  );

  const isNew = !hubEntry;

  const formRef = useRef<Form<Entry>>(null);

  const onFieldChanged: FormFieldChangeHandler<Entry> = (id, value, hasChanges) => {
    if (id.includes('hub_url')) {
      if (
        !isAdcHub(value.generic!.hub_url) &&
        value.share_profile !== ShareProfileConstants.HIDDEN_PROFILE_ID
      ) {
        // Reset share profile
        return Promise.resolve({
          share_profile: null,
        });
      }
    }

    return null;
  };

  const handleSave = () => {
    return formRef.current!.save();
  };

  const onSave: FormSaveHandler<Entry> = (changedFields) => {
    const updatedEntryFields = toFavoriteHub(changedFields);
    if (isNew) {
      return socket.post(FavoriteHubConstants.HUBS_URL, updatedEntryFields);
    }

    return socket.patch(
      `${FavoriteHubConstants.HUBS_URL}/${hubEntry.id}`,
      updatedEntryFields,
    );
  };

  const onFieldSetting: FormFieldSettingHandler<Entry> = (
    id,
    fieldOptions,
    formValue,
  ) => {
    if (id === 'share_profile') {
      const hubUrl = formValue.generic ? formValue.generic.hub_url : undefined;

      // Since the options are added dynamically based on the URL protocol, they must be
      // normalized to use the tcomb format
      Object.assign(fieldOptions, {
        nullOption,
        factory: t.form.Select,
        options: getFieldProfiles(profiles, hubUrl, formatter),
        transformer: intTransformer,
      });
    } else if (id === 'connection_mode_v4' || id === 'connection_mode_v6') {
      Object.assign(fieldOptions, {
        nullOption,
        transformer: intTransformer,
      });
    } else if (id === 'use_main_chat_notify' || id === 'show_joins') {
      Object.assign(fieldOptions, {
        template: IndeterminateCheckboxField,
        transformer: {
          format: (value: boolean | null) => (value === undefined ? null : value),
          parse: (value: boolean | null) => value,
        },
      });
    }
  };

  const { translate } = favT;
  const title = translate(isNew ? 'Add favorite hub' : 'Edit favorite hub');
  return (
    <RouteModal
      className="fav-hub"
      title={title}
      onApprove={handleSave}
      closable={false}
      icon={IconConstants.FAVORITE}
    >
      <Form<Entry>
        ref={formRef}
        onFieldChanged={onFieldChanged}
        onFieldSetting={onFieldSetting}
        onSave={onSave}
        fieldDefinitions={formDefinitions}
        sourceValue={formValue}
        location={location}
      />
    </RouteModal>
  );
};

export default ModalRouteDecorator<FavoriteHubDialogProps>(
  ShareProfileDecorator(FavoriteHubDialog, true, {
    urls: {
      hubEntry: ({ params }, socket) => {
        if (!params.entryId) {
          return Promise.resolve(undefined);
        }

        return socket.get(`${FavoriteHubConstants.HUBS_URL}/${params.entryId}`);
      },
    },
  }),
  'entries/:entryId?',
);
