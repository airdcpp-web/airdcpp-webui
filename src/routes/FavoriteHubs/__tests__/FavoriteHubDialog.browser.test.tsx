import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import { useTranslation } from 'react-i18next';
import { waitFor } from '@testing-library/dom';

import { getConnectedSocket, getMockServer } from 'airdcpp-apisocket/tests';

import { renderDataRoutes } from '@/tests/render/test-renderers';

import * as UI from '@/types/ui';

import {
  createTestRouteModalController,
  TestRouteModalNavigateButton,
} from '@/tests/helpers/test-dialog-helpers';
import FavoriteHubDialog from '../components/FavoriteHubDialog';
import { getModuleT } from '@/utils/TranslationUtils';
import ShareProfileConstants from '@/constants/ShareProfileConstants';
import { ShareProfilesListResponse } from '@/tests/mocks/api/share-profiles';
import FavoriteHubConstants from '@/constants/FavoriteHubConstants';
import {
  MOCK_FAVORITE_HUB_ID,
  FavoriteHubGetResponse,
} from '@/tests/mocks/api/favorite-hubs';
import { setInputFieldValues, setupUserEvent } from '@/tests/helpers/test-form-helpers';

// tslint:disable:no-empty
describe('FavoriteHubDialog', () => {
  let server: ReturnType<typeof getMockServer>;
  const getSocket = async () => {
    const { socket } = await getConnectedSocket(server);

    // Data fetch
    server.addRequestHandler(
      'GET',
      ShareProfileConstants.PROFILES_URL,
      ShareProfilesListResponse,
    );
    server.addRequestHandler(
      'GET',
      `${FavoriteHubConstants.HUBS_URL}/${MOCK_FAVORITE_HUB_ID}`,
      FavoriteHubGetResponse,
    );

    // Listeners
    server.addSubscriptionHandler(
      ShareProfileConstants.PROFILES_URL,
      'share_profile_added',
      undefined,
    );
    server.addSubscriptionHandler(
      ShareProfileConstants.PROFILES_URL,
      'share_profile_updated',
      undefined,
    );
    server.addSubscriptionHandler(
      ShareProfileConstants.PROFILES_URL,
      'share_profile_removed',
      undefined,
    );

    const onCreated = vi.fn();
    const onUpdated = vi.fn();

    // Save handlers
    server.addRequestHandler(
      'PATCH',
      `${FavoriteHubConstants.HUBS_URL}/${MOCK_FAVORITE_HUB_ID}`,
      undefined,
      onUpdated,
    );
    server.addRequestHandler('POST', FavoriteHubConstants.HUBS_URL, undefined, onCreated);

    return { socket, server, onCreated, onUpdated };
  };

  const renderDialog = async (id: number | null) => {
    const { socket, server, ...other } = await getSocket();

    const FavoriteHubDialogTest = () => {
      const { t } = useTranslation();
      const favT = getModuleT(t, UI.Modules.FAVORITE_HUBS);
      return (
        <>
          <TestRouteModalNavigateButton
            modalRoute={id ? `/home/entries/${id}` : '/home/entries'}
          />
          <FavoriteHubDialog favT={favT} />
        </>
      );
    };

    const routes = [
      {
        path: '/home/*',
        Component: FavoriteHubDialogTest,
      },
    ];

    const renderData = renderDataRoutes(routes, {
      socket,
      routerProps: { initialEntries: ['/home'] },
    });

    const modalController = createTestRouteModalController(renderData);
    return { modalController, ...renderData, ...other };
  };

  beforeEach(() => {
    server = getMockServer();
  });

  afterEach(() => {
    server.stop();
  });

  test('should update existing', async () => {
    const userEvent = setupUserEvent();
    const { getByText, getByLabelText, modalController, onUpdated } =
      await renderDialog(MOCK_FAVORITE_HUB_ID);

    await modalController.openDialog();

    // Check content
    await waitFor(() => expect(getByText('Edit favorite hub')).toBeTruthy());

    // Edit
    await setInputFieldValues(
      { userEvent, getByLabelText },
      {
        Name: 'Updated name',
      },
    );

    await modalController.closeDialogButton('Save');

    expect(onUpdated).toHaveBeenCalledTimes(1);
    expect(onUpdated.mock.calls[0]).toMatchSnapshot();
  });

  test('should create new', async () => {
    const userEvent = setupUserEvent();
    const { getByText, getByLabelText, modalController, onCreated } =
      await renderDialog(null);

    await modalController.openDialog();

    // Check content
    await waitFor(() => expect(getByText('Add favorite hub')).toBeTruthy());

    // Edit
    await setInputFieldValues(
      { userEvent, getByLabelText },
      {
        Name: 'Hub name',
        'Hub URL': 'adcs://example.com:1511 ',
      },
    );

    await modalController.closeDialogButton('Save');

    expect(onCreated).toHaveBeenCalledTimes(1);
    expect(onCreated.mock.calls[0]).toMatchSnapshot();
  });
});
