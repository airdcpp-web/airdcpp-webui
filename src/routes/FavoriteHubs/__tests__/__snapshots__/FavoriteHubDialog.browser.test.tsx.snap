// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FavoriteHubDialog > should create new 1`] = `
{
  "data": {
    "auto_connect": false,
    "away_message": null,
    "connection_ip_v4": null,
    "connection_ip_v6": null,
    "connection_mode_v4": null,
    "connection_mode_v6": null,
    "hub_description": null,
    "hub_url": "adcs://example.com:1511",
    "name": "Hub name",
    "nick": null,
    "share_profile": null,
    "show_joins": null,
    "use_main_chat_notify": null,
    "user_description": null,
  },
  "method": "POST",
  "path": "favorite_hubs",
}
`;

exports[`FavoriteHubDialog > should update existing 1`] = `
{
  "data": {
    "name": "Updated name",
  },
  "method": "PATCH",
  "path": "favorite_hubs/703302715",
}
`;
